In [1]:
import sys
import os
sys.path.append('/content/drive/MyDrive/Kernel_Challenge/')
from tqdm.notebook import tqdm as tqdm
from kernel import *
from classifiers import *
from utils.utils import return_training_datasets, split_train_val

In [ ]:
all_x, all_y = return_training_datasets('/content/drive/MyDrive/Kernel_Challenge/kernel_data')

# Xtr0

In [3]:
X, y = all_x[0], all_y[0]
x_train, x_val, y_train, y_val = split_train_val(X,y)

## Spectrum Kernel Grid Search

In [5]:
k_candidates = [5, 6, 7, 8, 9, 10, 11, 12]
matrix_ready = {k: False for k in k_candidates}

### SVM

In [15]:
C_candidates = [1, 0.5, 0.1, 0.07, 0.03, 0.01]

In [ ]:
summary_SKSVM = dict()
for k in tqdm(k_candidates):
  for C in tqdm(C_candidates):
    Kernel = SpectrumKernel(k=k, add_inverse=True)
    clf = SVM(Kernel, C=C, save_dir='/content/drive/MyDrive/Kernel_Challenge/kernel_data/', version="Xtr0")
    path_2_save = os.path.join(clf.save_dir, clf.kernel.name + f"_{clf.version}" + ".pkl")
    if matrix_ready[k]:
      kernel = Kernel.load_kernel(path_2_save)
      clf.kernel = kernel
      clf._fit(y_train.Bound)
    else:
      clf.fit(x_train, y_train.Bound)
      matrix_ready[k] = True
    y_pred = clf.predict(x_val)
    score = sum(y_pred==y_val.Bound)/len(y_pred)
    summary_SKSVM[f"C={C} and k={k} : Accuracy on validation"] = score

In [14]:
summary_SKVM

{'C=0.007 and k=3 : Accuracy on validation': 0.5925,
 'C=0.007 and k=4 : Accuracy on validation': 0.615,
 'C=0.007 and k=5 : Accuracy on validation': 0.6375,
 'C=0.007 and k=6 : Accuracy on validation': 0.6325,
 'C=0.007 and k=7 : Accuracy on validation': 0.65,
 'C=0.007 and k=8 : Accuracy on validation': 0.64,
 'C=0.01 and k=3 : Accuracy on validation': 0.585,
 'C=0.01 and k=4 : Accuracy on validation': 0.6125,
 'C=0.01 and k=5 : Accuracy on validation': 0.62,
 'C=0.01 and k=6 : Accuracy on validation': 0.65,
 'C=0.01 and k=7 : Accuracy on validation': 0.6425,
 'C=0.01 and k=8 : Accuracy on validation': 0.6525,
 'C=0.03 and k=3 : Accuracy on validation': 0.595,
 'C=0.03 and k=4 : Accuracy on validation': 0.6125,
 'C=0.03 and k=5 : Accuracy on validation': 0.6325,
 'C=0.03 and k=6 : Accuracy on validation': 0.6575,
 'C=0.03 and k=7 : Accuracy on validation': 0.6425,
 'C=0.03 and k=8 : Accuracy on validation': 0.645,
 'C=0.07 and k=3 : Accuracy on validation': 0.58,
 'C=0.07 and k=4 : A

In [16]:
k_candidates = [9, 10, 11, 12]
matrix_ready = {k: False for k in k_candidates}

In [ ]:
summary_SKSVM_2 = dict()
for k in tqdm(k_candidates):
  for C in tqdm(C_candidates):
    Kernel = SpectrumKernel(k=k, add_inverse=True)
    clf = SVM(Kernel, C=C, save_dir='/content/drive/MyDrive/Kernel_Challenge/kernel_data/', version="Xtr0")
    path_2_save = os.path.join(clf.save_dir, clf.kernel.name + f"_{clf.version}" + ".pkl")
    if matrix_ready[k]:
      kernel = Kernel.load_kernel(path_2_save)
      clf.kernel = kernel
      clf._fit(y_train.Bound)
    else:
      clf.fit(x_train, y_train.Bound)
      matrix_ready[k] = True
    y_pred = clf.predict(x_val)
    score = sum(y_pred==y_val.Bound)/len(y_pred)
    summary_SKSVM_2[f"C={C} and k={k} : Accuracy on validation"] = score

In [18]:
summary_SKSVM_2

{'C=0.01 and k=10 : Accuracy on validation': 0.66,
 'C=0.01 and k=11 : Accuracy on validation': 0.665,
 'C=0.01 and k=12 : Accuracy on validation': 0.63,
 'C=0.01 and k=9 : Accuracy on validation': 0.6525,
 'C=0.03 and k=10 : Accuracy on validation': 0.6675,
 'C=0.03 and k=11 : Accuracy on validation': 0.6625,
 'C=0.03 and k=12 : Accuracy on validation': 0.635,
 'C=0.03 and k=9 : Accuracy on validation': 0.6425,
 'C=0.07 and k=10 : Accuracy on validation': 0.67,
 'C=0.07 and k=11 : Accuracy on validation': 0.6625,
 'C=0.07 and k=12 : Accuracy on validation': 0.635,
 'C=0.07 and k=9 : Accuracy on validation': 0.6425,
 'C=0.1 and k=10 : Accuracy on validation': 0.67,
 'C=0.1 and k=11 : Accuracy on validation': 0.6625,
 'C=0.1 and k=12 : Accuracy on validation': 0.635,
 'C=0.1 and k=9 : Accuracy on validation': 0.6425,
 'C=0.5 and k=10 : Accuracy on validation': 0.67,
 'C=0.5 and k=11 : Accuracy on validation': 0.6625,
 'C=0.5 and k=12 : Accuracy on validation': 0.635,
 'C=0.5 and k=9 : A

Summary : \
C=0.1 and k=10 : Accuracy on validation': 0.67 \
C=0.01 and k=11 : Accuracy on validation': 0.665 \


### Kernel Logisitic Regression

In [19]:
k_candidates = [5, 6, 7, 8, 9, 10, 11, 12]
matrix_ready = {k: True for k in k_candidates}

In [20]:
lambda_candidates = [0.1, 0.05, 0.01, 0.005, 0.001, 0.0005, 0.0001]

In [ ]:
summary_SKLR = dict()
for k in tqdm(k_candidates):
  for lmbd in tqdm(lambda_candidates):
    Kernel = SpectrumKernel(k=k, add_inverse=True)
    clf = KernelLogisticRegression(Kernel, lambda_=lmbd, save_dir='/content/drive/MyDrive/Kernel_Challenge/kernel_data/', max_iter=10000, version="Xtr0")
    path_2_save = os.path.join(clf.save_dir, clf.kernel.name + f"_{clf.version}" + ".pkl")
    if matrix_ready[k]:
      kernel = Kernel.load_kernel(path_2_save)
      clf.kernel = kernel
      clf._fit(y_train.Bound)
    else:
      clf.fit(x_train, y_train.Bound)
      matrix_ready[k] = True
    y_pred = clf.predict(x_val)
    score = sum(y_pred==y_val.Bound)/len(y_pred)
    summary_SKLR[f"lambda={lmbd} and k={k} : Accuracy on validation"] = score

In [22]:
summary_SKLR

{'lambda=0.0001 and k=10 : Accuracy on validation': 0.6525,
 'lambda=0.0001 and k=11 : Accuracy on validation': 0.655,
 'lambda=0.0001 and k=12 : Accuracy on validation': 0.63,
 'lambda=0.0001 and k=5 : Accuracy on validation': 0.5775,
 'lambda=0.0001 and k=6 : Accuracy on validation': 0.6325,
 'lambda=0.0001 and k=7 : Accuracy on validation': 0.6375,
 'lambda=0.0001 and k=8 : Accuracy on validation': 0.6475,
 'lambda=0.0001 and k=9 : Accuracy on validation': 0.6475,
 'lambda=0.0005 and k=10 : Accuracy on validation': 0.6575,
 'lambda=0.0005 and k=11 : Accuracy on validation': 0.645,
 'lambda=0.0005 and k=12 : Accuracy on validation': 0.6275,
 'lambda=0.0005 and k=5 : Accuracy on validation': 0.5925,
 'lambda=0.0005 and k=6 : Accuracy on validation': 0.62,
 'lambda=0.0005 and k=7 : Accuracy on validation': 0.65,
 'lambda=0.0005 and k=8 : Accuracy on validation': 0.665,
 'lambda=0.0005 and k=9 : Accuracy on validation': 0.635,
 'lambda=0.001 and k=10 : Accuracy on validation': 0.665,
 '

Summary : \
lambda=0.005 and k=11 : Accuracy on validation': 0.675 \
lambda=0.01 and k=7 : Accuracy on validation': 0.6675 \

## Substring Kernel Grid Search

In [10]:
k_candidates = [6, 7, 8, 9, 10, 11]
m_candidates = [1, 2, 3]
matrix_ready = {(k,m): False for k,m in zip(k_candidates, m_candidates)}

### SVM

In [11]:
C_candidates = [1, 0.1, 0.01]

In [17]:
summary_SubKSVM = dict()
for k in tqdm(k_candidates):
  for m in tqdm(m_candidates):
    for C in tqdm(C_candidates):
      Kernel = SubstringKernel(k=k, delta=1, m=m)
      clf = SVM(Kernel, C=C, save_dir='/content/drive/MyDrive/Kernel_Challenge/kernel_data/', version="Xtr0")
      path_2_save = os.path.join(clf.save_dir, clf.kernel.name + f"_{clf.version}" + ".pkl")
      if matrix_ready[(k,m)]:
        kernel = Kernel.load_kernel(path_2_save)
        clf.kernel = kernel
        clf._fit(y_train.Bound)
      else:
        clf.fit(x_train, y_train.Bound)
        matrix_ready[k] = True
      y_pred = clf.predict(x_val)
      score = sum(y_pred==y_val.Bound)/len(y_pred)
      summary_SubKSVM[f"C={C} and k={k} and m={m} : Accuracy on validation"] = score

Gram Matrix Built.


ValueError: ignored

Summary : \
C=0.1 and k=10 : Accuracy on validation': 0.67 \
C=0.01 and k=11 : Accuracy on validation': 0.665 \


### Kernel Logisitic Regression

In [ ]:
k_candidates = [6, 7, 8, 9, 10, 11]
m_candidates = [1, 2, 3]
matrix_ready = {(k,m): True for k,m in zip(k_candidates, m_candidates)}

In [23]:
lambda_candidates = [0.01, 0.005, 0.001]

In [ ]:
summary_SubKLR = dict()
for k in tqdm(k_candidates):
  for lmbd in tqdm(lambda_candidates):
    Kernel = SpectrumKernel(k=k, add_inverse=True)
    clf = KernelLogisticRegression(Kernel, lambda_=lmbd, save_dir='/content/drive/MyDrive/Kernel_Challenge/kernel_data/', max_iter=10000, version="Xtr0")
    path_2_save = os.path.join(clf.save_dir, clf.kernel.name + f"_{clf.version}" + ".pkl")
    if matrix_ready[k]:
      kernel = Kernel.load_kernel(path_2_save)
      clf.kernel = kernel
      clf._fit(y_train.Bound)
    else:
      clf.fit(x_train, y_train.Bound)
      matrix_ready[k] = True
    y_pred = clf.predict(x_val)
    score = sum(y_pred==y_val.Bound)/len(y_pred)
    summary_SubKLR[f"lambda={lmbd} and k={k} and m={m} : Accuracy on validation"] = score

In [ ]:
summary_SKLR

{'lambda=0.0001 and k=10 : Accuracy on validation': 0.6525,
 'lambda=0.0001 and k=11 : Accuracy on validation': 0.655,
 'lambda=0.0001 and k=12 : Accuracy on validation': 0.63,
 'lambda=0.0001 and k=5 : Accuracy on validation': 0.5775,
 'lambda=0.0001 and k=6 : Accuracy on validation': 0.6325,
 'lambda=0.0001 and k=7 : Accuracy on validation': 0.6375,
 'lambda=0.0001 and k=8 : Accuracy on validation': 0.6475,
 'lambda=0.0001 and k=9 : Accuracy on validation': 0.6475,
 'lambda=0.0005 and k=10 : Accuracy on validation': 0.6575,
 'lambda=0.0005 and k=11 : Accuracy on validation': 0.645,
 'lambda=0.0005 and k=12 : Accuracy on validation': 0.6275,
 'lambda=0.0005 and k=5 : Accuracy on validation': 0.5925,
 'lambda=0.0005 and k=6 : Accuracy on validation': 0.62,
 'lambda=0.0005 and k=7 : Accuracy on validation': 0.65,
 'lambda=0.0005 and k=8 : Accuracy on validation': 0.665,
 'lambda=0.0005 and k=9 : Accuracy on validation': 0.635,
 'lambda=0.001 and k=10 : Accuracy on validation': 0.665,
 '

Summary : \
lambda=0.005 and k=11 : Accuracy on validation': 0.675 \
lambda=0.01 and k=7 : Accuracy on validation': 0.6675 \